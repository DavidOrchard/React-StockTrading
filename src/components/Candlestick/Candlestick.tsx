import React, { useMemo } from 'react';

// point contains [lowestBid, lowestSale, highestSale, highestAsk
// they are 100 apart
type Point = number[];
const WIDTH = 40;
const HEIGHT = 300;
const OFFSET = WIDTH * 2;

const makeStick = (point: Point, index: number) => {
    const [bottom, open, close, top] = point;
    return (
        <g key={index} aria-label={`candle number ${index}`}>
            <rect x={WIDTH + index * OFFSET - 1} y={HEIGHT - top} width="2" height={top-bottom} stroke="none" strokeWidth="0" fill="#3366cc"/>
            <rect x={WIDTH/2 + index * OFFSET} y={HEIGHT - open} width={WIDTH} height={Math.abs(close-open)} stroke="#3366cc" strokeWidth="2" fill={close-open > 0 ? "#3366cc" : "#ffffff"}/>
        </g>);

};

// Make SVG Candlestick based upon points passed in.
export const Candlestick = ({points}:{points: Point[]}) => useMemo(
    () => (
        <svg width="900" height="300">
            <rect x="0" y="0" width="900" height="300" stroke="none" strokeWidth="0" fill="#ffffff"></rect>
            {points.map(makeStick)}
        </svg>
    ),
    [points]
);
