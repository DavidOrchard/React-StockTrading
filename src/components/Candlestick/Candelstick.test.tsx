import React from 'react';
import { render, screen } from '@testing-library/react';
import { Candlestick } from './Candlestick';

test('renders Candlestick', () => {
  render(<Candlestick points={[[185, 218, 232, 260]]}/>);
  const candlestick = screen.getByLabelText(/candle number 0/i);
  expect(candlestick).toMatchSnapshot();
});