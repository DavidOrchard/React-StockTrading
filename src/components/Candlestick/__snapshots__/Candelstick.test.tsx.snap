// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Candlestick 1`] = `
<g
  aria-label="candle number 0"
>
  <rect
    fill="#3366cc"
    height="75"
    stroke="none"
    stroke-width="0"
    width="2"
    x="39"
    y="40"
  />
  <rect
    fill="#3366cc"
    height="14"
    stroke="#3366cc"
    stroke-width="2"
    width="40"
    x="20"
    y="82"
  />
</g>
`;
